frontend/rust-lib/flowy-database2/src/services/calculations/entities.rs

```diff
@@ -0,0 +1,248 @@
+          .get_i64_value(CALCULATION_TYPE)
+          .unwrap_or_default();
+
+        Ok(Calculation {
+          id,
+          field_id,
+          calculation_type,
+          value,
+        })
+      },
+      _ => {
+        bail!("Invalid calculation data")
+      },
+    }
+  }
+}
+
+pub struct CalculationsResultNotification {
+  pub view_id: String,
+
+  pub calculations: Vec<Calculation>,
+}
+
+impl CalculationsResultNotification {
+  pub fn new(view_id: String) -> Self {
+    Self {
+      view_id,
+      calculations: vec![],
+    }
+  }
+}
+
+impl Calculation {
+  pub fn none(id: String, field_id: String) -> Self {
+    Self {
+      id,
+      field_id,
+      calculation_type: 0,
+      value: "".to_owned(),
+    }
+  }
+
+  pub fn calculate(&self, row_cells: Vec<Arc<RowCell>>) -> String {
+    let ty: CalculationType = self.calculation_type.into();
+
+    match ty {
+      CalculationType::Average => self.calculate_average(row_cells),
+      CalculationType::Max => self.calculate_max(row_cells),
+      CalculationType::Median => self.calculate_median(row_cells),
+      CalculationType::Min => self.calculate_min(row_cells),
+      CalculationType::Sum => self.calculate_sum(row_cells),
+    }
+  }
+
+  fn calculate_average(&self, row_cells: Vec<Arc<RowCell>>) -> String {
+    let mut sum = 0.0;
+    let mut len = 0.0;
+    for row_cell in row_cells {
+      if let Some(cell) = &row_cell.cell {
+        let data = cell.get("data");
+        if let Some(data) = data {
+          match data.to_string().parse::<f64>() {
+            Ok(value) => {
+              sum += value;
+              len += 1.0;
+            },
+            _ => tracing::info!("Failed to parse ({}) to f64", data),
+          }
+        }
+      }
+    }
+
+    if len > 0.0 {
+      format!("{:.5}", sum / len)
+    } else {
+      "0".to_owned()
+    }
+  }
+
+  fn calculate_median(&self, row_cells: Vec<Arc<RowCell>>) -> String {
+    let values = self.reduce_values_f64(row_cells, |values| {
+      values.sort_by(|a, b| a.partial_cmp(b).unwrap());
+      values.clone()
+    });
+
+    if !values.is_empty() {
+      Self::median(&values).to_string()
+    } else {
+      "".to_owned()
+    }
+  }
+
+  fn median(array: &Vec<f64>) -> f64 {
+    if (array.len() % 2) == 0 {
+      let left = array.len() / 2 - 1;
+      let right = array.len() / 2;
+      (array[left] + array[right]) / 2.0
+    } else {
+      array[array.len() / 2]
+    }
+  }
+
+  fn calculate_min(&self, row_cells: Vec<Arc<RowCell>>) -> String {
+    let values = self.reduce_values_f64(row_cells, |values| {
+      values.sort_by(|a, b| a.partial_cmp(b).unwrap());
+      values.clone()
+    });
+
+    if !values.is_empty() {
+      let min = values.iter().min_by(|a, b| a.total_cmp(b));
+      if let Some(min) = min {
+        return min.to_string();
+      }
+    }
+
+    "".to_owned()
+  }
+
+  fn calculate_max(&self, row_cells: Vec<Arc<RowCell>>) -> String {
```

Comment:
Entities should not include business logic; they are meant to be plain entities only. Consider relocating the calculation logic closer to the CalculationController.